### ISSUE TITLE
Doxygen documentation

### ISSUE THREAD

**Issue by benkrikler** [alcap-org/AlcapDAQ#54]
In anticipation of people using rootana more and more, I'm setting up doxygen to automatically regenerate documentation each night.  

There are several places I can host it:
1. Here on this repository.  
   - Pro: Everyone can see it and github issues can reference commits easily
   - Con: Messes up network diagram
2. Make a pages repository for the organisation which we could host this and other web stuff in
   - Pro: Clean and everyone can see it
   - Con: New repository and broken away from source
3. Host it on my machine at Imperial
   - Pro: Quite a bit simpler.  Already done for practise (http://www.hep.ph.ic.ac.uk/~bek07/alcap/AlcapDAQ_documentation/rootana/html/annotated.html)
   - Con: Not very egalitarian, no versioning of documentation

Any preferences or opinions?

**Comment by AndrewEdmonds11** [comment]
I really like looking at the network diagram so I would prefer either option 2 or 3.

**Comment by jrquirk** [comment]
Can you explain or provide a link to explain 2 a bit more? I like the sounds of it as well.

**Comment by benkrikler** [comment]
This link does a good job at describing option 1 and 2 together: https://help.github.com/articles/user-organization-and-project-pages

I've used the organisation page before and it works well.

**Comment by benkrikler** [comment]
If no one objects I'm going to use point 2 above and set up a new repository for doxygen and possibly other web-stuff in the future.

**Comment by jrquirk** [comment]
I like it. Where are the doxygen files stored? Will all of those files be stored only on the new alcap-org.github.io repo, or will they be in AlcapDAQ and then save the output to the new repo? And in either case, I'd like to take a look at them to help figure out how doxygen works; I've never used it on a project before.

At this point it's only implemented for rootana, but in the not-so-distant future it'd be nice to use it for the other programs we use (alcapana for instance, the ODBCheck program, the picture-book stuff and other scripts; though the scope of doxygen is unknown to me so some of this might not work). So maybe there should be an announcement that it's time people started familiarizing themselves with it. Especially before people start developing in rootana.

NB: The rootana code in develop doesn't seem to have doxygen comments. Actually, it seems to have few if any comments, and this seems like as good a time as any to point that out.

**Comment by benkrikler** [comment]
> Where are the doxygen files stored?

Do you mean the files doxygen uses to create the html documentation or the html documentation itself?  If its the first one, there is likely to be at most 2 files used by doxygen to generate the documentation, aside from the c++ code base itself.  One of those files is a configuration file, which I have already made and is sitting in rootana/doxygen on develop.  The second possible file is a main-page which would be used to create the landing page, although I'm not sure if we would need this as we already have the wiki which is much nicer to work with.

> At this point it's only implemented for rootana,

Actually, although I've put the config file in rootana, there's nothing to stop it being reused for alcapana directly.  I could in principle run over the entire AlcapDAQ directory and put up a single set of documentation for it all. I can tell it to skip code in the midas and root directories for example.
It could also help with some of the scripts we have, though I have no idea how well this works; I know doxygen was originally written for c / c++ but has since been extended to other languages.  I'm not sure how much we need that though.

My priority now is to get rootana documented and this is one tool that is able to be maintained automatically.  This needs rootana to be well commented and as you point out that's not the case right now.  I plan to go through all headers and comment it all very shortly, but if I do that before these interfaces are fixed (ie. once the event navigator is ready), I fear I'd be doubling my work load.

**Comment by benkrikler** [comment]
New repository is up and I've added a basic directory structure.  Here's how things should look when finished:

```
alcap-org.github.io
  \---doxygen
        ¦---build_files
        ¦        ¦----Doxyconfig
        ¦        \----run_doxygen.sh
        \---html
                \----all the different files produced by doxygen
```

**Comment by benkrikler** [comment]
This is fully up and running now and the code is browseable at:
http://alcap-org.github.io/doxygen/AlcapDAQ/develop/annotated.html

We can tweak the doxygen configuration as we like, @jrquirk feel free to swap in the file you were setting up.  I am running this on all of AlcapDAQ, not just rootana.  I'm skipping ROOT, but I've kept in Midas.  I think that might be a mistake though as it adds a lot of other stuff to the documentation and I'm not sure if it's useful.  What do people think?

This will be automatically updated (I hope) every night so it should track the develop branch with no problem.  **Please don't commit to the doxygen directory on the new repository without telling me as that could lead to conflicts which would break the automation.**

**Comment by AndrewEdmonds11** [comment]
Thanks, Ben. So what do the rest of us need to do? Do we need to go back and make proper doxygen comments?

**Comment by benkrikler** [comment]
If we want to really make the most of this, then everything we write should have proper doxygen comments.  Normally that just means precede every function, data member or class declaration with triple backslash comments.  These get incorporated into the html as descriptions of each item.  You can do much fancier stuff like build Todo lists, add deprecated warnings etc, but I don't expect we'll use this much.  Here's a cheatsheet that might help: http://www.agapow.net/programming/doxygen-cheatsheet-2

Even  if we don't do this, doxygen can still be pretty helpful.  If you want to know what calls a given method or what a given method uses, as well as class inheritance and so on, then doxygen does that for us already, and we don't need extra comments.

So to summarise:  if we go back and comment things properly, it would be very helpful but not essential.

**Comment by litchfld** [comment]
> I've kept in Midas. I think that might be a mistake though as it adds a lot of other stuff to the documentation and I'm not sure if it's useful. What do people think?

Ugh, get rid of it. If you could put it up as a separate tree for reference, that would be fine, but as it is it's mostly horrible C clutter.

**Comment by benkrikler** [comment]
> Ugh, get rid of it

Done.  

I'm tempted to remove all directories accept for AlcapDAQ/analyzer and ignore the md5-lib and compress directories in that.  I think that would cover the main code from alcapana, rootana and odb-check in one go.

**Comment by jrquirk** [comment]
It'd be nice to find a way to break this up by project.

The only common codes are TGlobalData, TSetupData, and TPulseIsland between alcapana and rootana. And I don't think people want to see the [ODBCheck](http://alcap-org.github.io/doxygen/AlcapDAQ/develop/classODBCheck.html) code when looking for info on ExportPulse.

**Comment by benkrikler** [comment]
Ok, we're down to just rootana and analyzer/src/common.  If we want to build other projects in parallel, hopefully the script would require minimal effort to have it do this as well.

I'm going to close this now, but feel free to add more comments and reopen it if you think we need.
